### ISSUE TITLE
docker_update.sh results in unconnectable database

### ISSUE THREAD

**Issue by treyturner** [AzuraCast/AzuraCast#216]
I've come across this twice now and haven't yet figured out how to resolve. I start with a working instance and try to either run docker_update.sh or the manual commands specified in README.md to pull updates. The pulls seem to go fine, but in the playbook I get:

```
Creating network "azuracast_default" with the default driver
Creating azuracast_redis_1 ...
Creating azuracast_influxdb_1 ...
Creating azuracast_stations ...
Creating azuracast_redis_1
Creating azuracast_mariadb_1 ...
Creating azuracast_influxdb_1
Creating azuracast_mariadb_1
Creating azuracast_stations ... done
Creating azuracast_web_1 ...
Creating azuracast_web_1 ... done
Updating AzuraCast (Environment: production)
HEAD is now at 59f8075 #215 -- Fix namespacing issue with deleting playlists.
remote: Counting objects: 20, done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 20 (delta 15), reused 18 (delta 13), pack-reused 0
Unpacking objects: 100% (20/20), done.
From https://github.com/AzuraCast/AzuraCast
   59f8075..cbefa9e  master     -> origin/master
Updating 59f8075..cbefa9e
Fast-forward
 app/config/forms/user.conf.php                 |  8 +++
 app/src/AzuraCast/Radio/Frontend/IceCast.php   | 15 ++++--
 tests/functional/B01_Admin_ApiKeysCest.php     |  7 ++-
 tests/functional/B02_Admin_RecordsCest.php     | 68 ++++++++++++++++++++++----
 tests/functional/C03_Station_PlaylistsCest.php |  6 +++
 tests/functional/CestAbstract.php              |  1 +
 6 files changed, 90 insertions(+), 15 deletions(-)

PLAY [all] *********************************************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [azuracast-update : Make update script executable again] ******************
changed: [localhost]

TASK [azuracast-update : Install Composer Dependencies (Development)] **********
skipping: [localhost]

TASK [azuracast-update : Install Composer Dependencies (Production)] ***********
changed: [localhost]

TASK [azuracast-update : Update configuration from legacy] *********************
changed: [localhost]

TASK [azuracast-update : Clear AzuraCast Cache] ********************************
changed: [localhost]

TASK [azuracast-update : Update Database] **************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "php /var/azuracast/www/                                                                                           util/cli.php migrations:migrate --no-interaction --allow-no-migration", "delta":                                                                                            "0:00:00.852878", "end": "2017-08-20 17:58:16.807683", "failed": true, "rc": 1,                                                                                            "start": "2017-08-20 17:58:15.954805", "stderr": "\n                                                                                                                                                                         \n  [Doctrine\\DBAL\\Exceptio                                                                                           n\\ConnectionException]                               \n  An exception occurred                                                                                            in driver: SQLSTATE[HY000] [2002] Connection refused  \n                                                                                                                                                                         \n\n                                                                                                                                        \n  [Doctrine\\DBAL\\Driver\\PDOException]        \n  SQL                                                                                           STATE[HY000] [2002] Connection refused  \n                                                                                                                                        \n\n                                             \n  [PDOException]                                                                                                                        \n  SQLSTATE[HY000] [2002] Connection refused  \n                                                                                                                                        \n\nmigrations:migrate [--write-sql] [--dry                                                                                           -run] [--query-time] [--allow-no-migration] [--configuration [CONFIGURATION]] [-                                                                                           -db-configuration [DB-CONFIGURATION]] [--] [<version>]", "stderr_lines": ["", "                                                                                                                                                                         ",                                                                                            "  [Doctrine\\DBAL\\Exception\\ConnectionException]                                                                                                                          ", "  An exception occurred in driver: SQLSTATE[HY000] [2002] Connection refus                                                                                           ed  ", "                                                                                                                                                                         ", "", "                                             ", "  [Doctrine\\DBAL                                                                                           \\Driver\\PDOException]        ", "  SQLSTATE[HY000] [2002] Connection refused                                                                                             ", "                                             ", "", "                                                                                                                                        ", "  [PDOException]                             ", "  SQL                                                                                           STATE[HY000] [2002] Connection refused  ", "                                                                                                                                        ", "", "migrations:migrate [--write-sql] [--dry-run] [--query-time] [--                                                                                           allow-no-migration] [--configuration [CONFIGURATION]] [--db-configuration [DB-CO                                                                                           NFIGURATION]] [--] [<version>]"], "stdout": "Loading configuration from the inte                                                                                           gration code of your framework (setter).\n                                                                                                                                                               \n                    Doctrine Database Migrations                                                                                                               \n                                                                                                                                                               ", "stdout_lines": ["Loading configuration from the integration code o                                                                                           f your framework (setter).", "                                                                                                                                                               ", "                    Doctrine Database Migrations                                                                                                               ", "                                                                                                                                                               "]}
        to retry, use: --limit @/var/azuracast/www/util/ansible/docker_update.re                                                                                           try

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=4    unreachable=0    failed=1

azuracast_redis_1 is up-to-date
azuracast_stations is up-to-date
azuracast_mariadb_1 is up-to-date
azuracast_influxdb_1 is up-to-date
azuracast_web_1 is up-to-date
Creating azuracast_cron ...
Creating azuracast_cli_1 ...
Creating azuracast_nginx_1 ...
Creating azuracast_nginx_1
Creating azuracast_cron
Creating azuracast_nginx_1 ... done
```

The mariadb container is running though:
```
root@breaks:~/AzuraCast# docker ps
CONTAINER ID        IMAGE                                 COMMAND                  CREATED             STATUS              PORTS                                        NAMES
51d35668afcd        azuracast/azuracast_web:latest        "cron && tail -f /..."   3 minutes ago       Up 3 minutes        9000/tcp                                     azuracast_cron
8ee10739d07f        azuracast/azuracast_nginx:latest      "nginx -g 'daemon ..."   3 minutes ago       Up 3 minutes        0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp     azuracast_nginx_1
6f417c7aae54        azuracast/azuracast_web:latest        "/usr/bin/supervis..."   3 minutes ago       Up 3 minutes        9000/tcp                                     azuracast_web_1
a554a62734c2        azuracast/azuracast_stations:latest   "/usr/bin/supervis..."   3 minutes ago       Up 3 minutes        0.0.0.0:8000-8500->8000-8500/tcp, 9001/tcp   azuracast_stations
cf658fb9f861        mariadb:latest                        "docker-entrypoint..."   3 minutes ago       Up 3 minutes        3306/tcp                                     azuracast_mariadb_1
2a90d16f6178        influxdb:alpine                       "/entrypoint.sh in..."   3 minutes ago       Up 3 minutes        8086/tcp                                     azuracast_influxdb_1
8fb4fa918ce4        redis:alpine                          "docker-entrypoint..."   3 minutes ago       Up 3 minutes        6379/tcp                                     azuracast_redis_1
```
and in that container, mysqld seems to be running:
```
root@breaks:~/AzuraCast# docker exec -it azuracast_mariadb_1 ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
mysql        1  0.1 19.8 1499860 99460 ?       Ssl  17:58   0:01 mysqld
root        71  0.0  0.4  17500  2012 pts/0    Rs+  18:12   0:00 ps aux
```
Repeated attempts to run the updates result in the same errors, and the service as a whole naturally won't start and isn't accessible via web. I really appreciate your timely responses to updating the code and am seeking how to safely pull in changes without putting my running instance at risk. I'm happy to try any steps or produce additional debug/logs as needed. Thanks!

**Comment by treyturner** [comment]
Ok I did just find this in the mariadb container dmesg:
```
[117574.237590] Out of memory: Kill process 31580 (mysqld) score 196 or sacrifice child
[117574.240760] Killed process 31580 (mysqld) total-vm:1481268kB, anon-rss:97796kB, file-rss:0kB
```
Am going to try increasing resources.

**Comment by treyturner** [comment]
That seems to have done the trick. Apparently the 512mb Digital Ocean droplets aren't quite enough to perform migrations.

**Comment by SlvrEagle23** [comment]
@treyturner Ahhh, that's no good, though. The 512MB DO VPS has worked like a charm on traditional installs for quite some time, though the solution to a lot of issues has been giving the VPS a little swap space (since the DO boxes don't come with this).

The goal with AzuraCast is to run on extremely commodity hardware, and to that end I've had it running on Raspberry Pi's and the cheapest VPSes I can find. I'm reopening this to see if there's a way I can optimize the Docker update process to get around this limitation.

**Comment by treyturner** [comment]
It might be worth mentioning that on the DO 512mb Docker droplet, I had to add an ```/etc/docker/daemon.json``` file in order for installation to progress past ```docker-compose run --rm cli azuracast_install```. It contained:

```
{
    "userland-proxy": false
}

If I didn't do this, the VM would fork a docker-proxy instance for every port that gets forwarded, which brought the box to a crawl.

**Comment by SlvrEagle23** [comment]
@treyturner That definitely shouldn't be happening on the newest Docker version. Older versions had that issue, but that's why I started bundling the installer itself along with the Docker install tools, just to ensure more people had the newer build that should fix that.

On my DO instance (the Ubuntu Docker image running on 512MB), even with the newest Docker I did run into an OOM shutdown issue until I gave the machine a little room to breathe via:

```
fallocate -l 2G /swapfile
chmod 600 /swapfile
mkswap /swapfile
swapon /swapfile
echo "/swapfile   none    swap    sw    0   0" >> /etc/fstab
```

Once I did that, it resolved both the initial install OOM issues and the update issues as well, allowing everything to carry on smoothly.

I think other than that, there's probably not much we can do for Docker installs on that low-spec of hardware. The traditional install is closer to the metal and is a better steward of resources (albeit less portable), and is probably the way to go for something that size.

If you're seeing the same issue with Docker being stubborn and spinning up a new docker-proxy instance for every individual port (particularly the huge range forwarded in the `stations` container) feel free to open a new ticket with that.

**Comment by SlvrEagle23** [comment]
@treyturner After looking into it, I was able to verify that even the newest version of Docker CE hasn't fixed the "one docker-proxy process per port forwarded" issue, even for large ranges of ports. Back in '15, they said they wouldn't fix it because they were going to rework the network stack; there's been no significant update on it since then, and it basically wipes out huge swaths of RAM when large port ranges need to be forwarded. Must not be a priority for them!

I've updated the default docker-compose.yml file to only forward the requisite ports for the first 10 hosted stations and their inbound streamer ports. From my tests, this pulls the overall RAM usage down to a level where even the 512MB DO droplet can easily run it without issue. If anyone still needs the original large port range forwarded, there's a new `docker-compose.stationports.yml` file that can be added (i.e. `docker-compose -f docker-compose.yml -f docker-compose.stationports.yml up -d`) if needed.
